src/rkopenmdao/error_controller.py

```diff
@@ -0,0 +1,74 @@
+from dataclasses import dataclass
+
+from rkopenmdao.error_estimator import ErrorEstimator
+
+
+def _three_queue_controller(_list: list, val) -> list:
+    if _list is None:
+        _list = [val]
+    elif len(_list) < 3:
+        _list.append(val)
+    else:
+        _list.pop(0)
+        _list.append(val)
+    return _list
+
+
+@dataclass
+class ErrorController:
+    alpha: float
+    beta: float = 0
+    gamma: float = 0
+    a: float = 0
+    b: float = 0
+    name: str = "ErrorController"
+    local_error_norms: list = None
+    delta_time_steps: list = None
+
+    def _estimate_next_step_function(self,
+                                     tolerance: float = 1e-3,
+                                     safety_factor: float = 0.95):
+
+        delta_time_new = safety_factor * self.delta_time_steps[-1]
+        if 0 in self.local_error_norms:
+            return 1.0
+        delta_time_new *= (tolerance / self.local_error_norms[-1]) ** self.alpha
+        if len(self.local_error_norms) >= 2:
+            delta_time_new *= (self.local_error_norms[-2] / tolerance) ** self.beta
+            if len(self.local_error_norms) >= 3:
+                delta_time_new *= (tolerance / self.local_error_norms[-3]) ** self.gamma
+        if len(self.delta_time_steps) >= 2:
+            delta_time_new *= (self.delta_time_steps[-1] / self.delta_time_steps[-2]) ** self.a
+            if len(self.delta_time_steps) >= 3:
+                delta_time_new *= (self.delta_time_steps[-2] / self.delta_time_steps[-3]) ** self.b
+        return delta_time_new
+
+    def __call__(self,
+                 current_norm: float,
+                 delta_t: float,
+                 tol: float = 1e-3,
+                 safety_factor: float = 0.95,
+                 ):
+        """ Estimates next possible step size for a given state and embedded solution
+        and returns whether the next step size meets the tolerance.
+        """
+        self.local_error_norms = _three_queue_controller(self.local_error_norms, current_norm)
+        self.delta_time_steps = _three_queue_controller(self.delta_time_steps, delta_t)
```

Comment:
Instead of your `_three_queue_controller`, I would probably just use a numpy arrays of size three in which you shift around the values. If you initialize them correctly (with tolerance for the `self.local_error_norms`, and with the original delta_t for `self.delta_time_steps`), you can even get rid of the size checks in `_esitmate_next_step_function`.

Reply:
This is actually a smart idea! Do you suggest to also state the tol and safety_factor in initialization instead of using it in the call function?

Reply:
Yes. tol and safety_factor are probably constant, so passing at initialization seems more fitting.